"Heartbeat" comms infrastructure (including prompted updates)

amstone326 opened the pull request:
This PR adds a "Heartbeat Manager" which is tied to the session service and will, as long as the session remains active, periodically (currently 1x per day) send a request to the server for a "heartbeat" for that app. 

The request from mobile can include any fixed-length pieces of information that are relevant to communicating a "heartbeat" about the device's current state. [This doc](https://docs.google.com/a/dimagi.com/document/d/1aw_m4fsCcZOho2qZsuKeT6DhhPE4Ol0gb4jvPoTMpK0/edit?usp=sharing) outlines the few things that I'm already including, as well as a list of several other things we could potentially include. @ctsims, we should either chat through the things on that list, or you can just comment with thoughts/additions.

The response from HQ should include any information that is useful toward that app maintaining a "healthy" state on the current device (right now, the only thing I'm checking for is if there is an .apk and/or .ccz updates that the user should be prompted/forced to do).

**Notes/Questions:**
- The HQ work for this has not been done or even resourced yet, so right now I've been parsing a test response to test everything. The code to make the request is in there but not getting called yet. Once the HQ side is actually done, I can swap out the call to ` parseTestHeartbeatResponse()` in favor of `requestHeartbeat(currentUser)`, and test it.
- The current UI and conceptual model for the update piece (i.e. how I distinguish between forced vs. prompted updates, etc.) is just my best first guess. We should do some UAT with interested parties to get feedback on how it seems to them. (Update: I'm waiting on hearing back from Saket from this)
- Right now, the screen to prompt/force an update on the user is only shown once after a login, if it is appropriate to do so, and then not again until the user logs in again. It should probably be shown more often than this, but I wasn't sure on a good rule, so I haven't implemented anything for this yet. @ctsims, let me know if you had anything in mind for this. Otherwise I think this is a good question for UAT, as well. 

**Screenshots below -->**

amstone326 commented:
This screen will appear upon login if there is currently an .apk or. ccz update to prompt:
![screenshot_20170420-153547](https://cloud.githubusercontent.com/assets/3723143/25386618/4022bb34-29c8-11e7-9f90-36450d30840c.png)

(and if only 1 is present, then only that shows):
![screenshot_20170425-145159](https://cloud.githubusercontent.com/assets/3723143/25386662/690aa070-29c8-11e7-852d-af9baee8cb4f.png)

If 1 of the updates is "forced" (meaning a force-by date was set and it is past that date), then that one is emphasized:
![screenshot_20170425-145130](https://cloud.githubusercontent.com/assets/3723143/25386666/6d0d2f08-29c8-11e7-99a0-40c6c766b8f6.png)

amstone326 commented:
@damagatchi retest this please

wpride commented:
@amstone326 conflicts

wpride commented on app/src/org/commcare/UpdateToPrompt.java:
```diff
@@ -0,0 +1,158 @@
+                ApkVersion currentVersion = new ApkVersion(pi.versionName);
+                return currentVersion.compareTo(this.apkVersion) < 0;
+            } catch (PackageManager.NameNotFoundException e) {
+                System.out.println("Couldn't get current .apk version to compare with in " +
```
Should be logged?

wpride commented on app/src/org/commcare/UpdateToPrompt.java:
```diff
@@ -0,0 +1,158 @@
+                baos.close();
+                serializedStream.close();
+            } catch (IOException e) {
+                System.out.println("IO error encountered while serializing UpdateToPrompt: "
```
Worth logging?

wpride commented on app/src/org/commcare/UpdateToPrompt.java:
```diff
@@ -0,0 +1,158 @@
+        this.versionString = ExtUtil.readString(in);
+        this.isApkUpdate = ExtUtil.readBool(in);
+        this.hasForceByDate = ExtUtil.readBool(in);
+        if (hasForceByDate) {
```
The usual pattern here is using an `ExtWrapNullable` - any reason you decided to do the checking manually?

amstone326 replied:
Oh, I guess just because I've never used that and wasn't super aware of it. Will adjust.

wpride commented on app/src/org/commcare/UpdateToPrompt.java:
```diff
@@ -0,0 +1,158 @@
+ * Created by amstone326 on 4/13/17.
+ */
+
+public class UpdateToPrompt implements Externalizable {
```
Did you consider using two separate classes for apkVersion and ccVersion, possibly with a shared super? Not certain that it'd be cleaner but it does seem like there's a good amount of `if`/`else` checking and not that many shared paths.

amstone326 replied:
I actually didn't consider it at all. I think you're right that it would be marginally cleaner, but I think it probably doesn't have enough of a clear advantage to be worth the refactor right now. Let me know if you disagree, though.

wpride commented on app/src/org/commcare/UpdateToPrompt.java:
```diff
@@ -0,0 +1,158 @@
+    public static final String KEY_APK_UPDATE_TO_PROMPT = "apk-update-to-prompt";
+
+    private String versionString;
+    private int cczVersion;
```
Total NIT, but the `Z` in `ccz` just stands for "zip file" so I think ccVersion would be more accurate (though completely feel free to ignore)

amstone326 replied:
Lol. Fair but definitely not going to change it because ccVersion could easily be misinterpreted to mean "CommCare version" which is not what that value represents.

wpride commented on app/src/org/commcare/UpdateToPrompt.java:
```diff
@@ -0,0 +1,158 @@
+    private int cczVersion;
+    private ApkVersion apkVersion;
+    private Date forceByDate;
+    private boolean hasForceByDate;
```
Seems like you could get rid of the `hasFroceByDate` variable and instead check for `forceByDate` being null

wpride commented on app/src/org/commcare/activities/HomeScreenBaseActivity.java:
```diff
@@ -94,16 +96,16 @@
-    protected static final int ADVANCED_ACTIONS_ACTIVITY = 1024;
-    protected static final int CREATE_PIN = 16384;
-    protected static final int AUTHENTICATION_FOR_PIN = 32768;
+    protected static final int MAKE_REMOTE_POST = 5;
```
✅ on the ridiculous int cleanup

wpride commented on app/src/org/commcare/activities/PromptUpdateActivity.java:
```diff
@@ -0,0 +1,185 @@
+import org.javarosa.core.services.locale.Localization;
+
+/**
+ * Created by amstone326 on 4/19/17.
```
Big enough class to merit a comment IMO

wpride commented on app/src/org/commcare/CommCareHeartbeatManager.java:
```diff
@@ -0,0 +1,287 @@
+            String prefsKey = forApkUpdate ?
+                    UpdateToPrompt.KEY_APK_UPDATE_TO_PROMPT : UpdateToPrompt.KEY_CCZ_UPDATE_TO_PROMPT;
+            String serializedUpdate = currentApp.getAppPreferences().getString(prefsKey, "");
+            if (serializedUpdate != null && !"".equals(serializedUpdate)) {
```
lint tells me `serializedUpdate != null` will always be `true`

amstone326 replied:
right it is

wpride commented on app/src/org/commcare/CommCareHeartbeatManager.java:
```diff
@@ -0,0 +1,287 @@
+import java.util.TimerTask;
+
+/**
+ * Created by amstone326 on 4/13/17.
```
Could use a comment

wpride replied:
Otherwise, seems like this class is still anticipated some changes from HQ until its finished up so will wait until this happens before digging in.

amstone326 replied:
Makes sense, although I don't anticipate many changes to this (besides actually calling the real request method instead of parsing the test response). The code that's there obviously needs to be tested, but I'm fairly confident it's not going to need a lot of changes, if any.

ctsims commented on app/src/org/commcare/CommCareHeartbeatManager.java:
```diff
@@ -0,0 +1,309 @@
+        }
+
+        private void processErrorResponse(int responseCode) {
+            Logger.log(AndroidLogger.TYPE_ERROR_SERVER_COMMS,
```
We may end up processing a lot of these, so I think we should be cautious to avoid logging errors to our forensics lots unless we can explicitly rate limit them.

amstone326 replied:
why would we expect to get a lot of these?

ctsims replied:
I hadn't realized when I commented that we weren't doing the retry yet, can chat over voice. I think we expect one of these each time we fail to process a response, which will likely occur more often if we retry more proactively.

ctsims commented on app/src/org/commcare/UpdateToPrompt.java:
```diff
@@ -0,0 +1,154 @@
+
+    private void writeToPrefsObject(SharedPreferences prefs) {
+        ByteArrayOutputStream baos = null;
+        DataOutputStream serializedStream = null;
```
I think there's an existing Utility method for "Pass in an externalizable, get back an byte[]", I'd try to re-use that as the error handling here is quite unfortunate to end up replicating

amstone326 replied:
Ah cool, that's much better

ctsims commented on app/src/org/commcare/UpdateToPrompt.java:
```diff
@@ -0,0 +1,154 @@
+                prefs.edit().putString(
+                        isApkUpdate ? KEY_APK_UPDATE_TO_PROMPT : KEY_CCZ_UPDATE_TO_PROMPT,
+                        serializedString).commit();
+                baos.close();
```
You don't need to put these here if you have them in the finally() block below

amstone326 replied:
removed

ctsims commented on app/src/org/commcare/CommCareHeartbeatManager.java:
```diff
@@ -0,0 +1,309 @@
+                    byte[] updateBytes = Base64.decode(serializedUpdate, Base64.DEFAULT);
+                    DataInputStream stream = new DataInputStream(new ByteArrayInputStream(updateBytes));
+                    UpdateToPrompt update = (UpdateToPrompt)
+                            ExtUtil.read(stream, UpdateToPrompt.class, ExtUtil.defaultPrototypes());
```
We have a different "master prototypes factory" set in Android that you can get from the context.

amstone326 replied:
(no longer relevant since I'm using the `SerializationUtil` method for this

ctsims commented on app/src/org/commcare/CommCareHeartbeatManager.java:
```diff
@@ -0,0 +1,309 @@
+                    }
+                } catch (Exception e) {
+                    // Something went wrong, so clear out whatever is there
+                    Logger.log(AndroidLogger.TYPE_ERROR_WORKFLOW,
```
I agree that crashing here in general if you can't read the update prompt is the right plan, but I'd wrap this call over the extutil.read() call specifically, rather than generically catching all errors

amstone326 replied:
Why? No matter what, we know that we’ve failed to read out whatever is currently stored in prefs, so we may as well clear it out or else we’re just going to encounter the same error over and over

ctsims replied:
Ah, I think I was unclear in this comment, the issue was't around why deserialization failed, it was that there was more code wrapped in the try/catch than just the read/deserialize, so in theory there could be other bugs in the code that instead of being properly thrown would get obfuscated as deserialization errors.

ctsims commented on app/src/org/commcare/CommCareApplication.java:
```diff
@@ -712,6 +712,7 @@ public void onServiceConnected(ComponentName className, IBinder service) {
                         }
                     }
 
+                    CommCareHeartbeatManager.instance().startHeartbeatCommunications();
```
I think the HeartbeatManager's lifecycle can be managed entirely within the CommCareService, which would remove the need for the singleton implementation. The CommCareService already manages its lifecycle cleanly, so if it handled these methods inside of the service ("one Heartbeat per session service") it would be equivalent

amstone326 replied:
"The CommCareService already manages its lifecycle cleanly," --> But this `onServiceConnected` method is clearly responsible for initializing a lot of things that are supposed to be tied to the start of the session. What makes this different?

ctsims commented on app/src/org/commcare/CommCareHeartbeatManager.java:
```diff
@@ -0,0 +1,309 @@
+        };
+
+        heartbeatTimer = new Timer();
+        heartbeatTimer.schedule(heartbeatTimerTask, new Date(), ONE_DAY_IN_MS);
```
The scheduling doesn't quite match my expectations. based on what we'd discussed previously, I think I would expect this process to be

[Manager] -> [One Thread Ever]

Manager starts the thread and wraps its events/lifecycle.

one- thread-ever is in a while() loop, that exits If and When [The thread succeeds] OR it reads a "stopPolling" flag 

stopHeartBeat -> one-thread-ever.stopPolling = true;

at the end of the while loop, if I/O was unsuccesful, the thead sleeps for a while(60 seconds)

When the thread wakes up it checks a flag that says "you should try again right now", if that's set to true, it tries again, if not it keeps sleeping in 60 second increments until some longer backoff (minutes/hours)

When a form push succeeds,the app can say "session.tellTheHeartbeatToTry->Manager->the-one-thread.tryagainflag = true"

amstone326 replied:
Hey, I'm a little confused about this comment. In the minimal discussions we had about how to set this up, you had explicitly said that you weren't sure whether it was preferable to tie the heartbeat's lifecycle to being triggered post-form submissions, vs. just set it up to happen on a timer, and that I should go with whichever choice ended up seeming easier/better to me. We also then later discussed that I was planning on using the timer strategy because I mentioned to you that I had found discussions online where people were wary of using a `TimerTask`, and you checked those out and then we agreed that it seemed like the concerns weren't relevant to us, and you agreed that that seemed like a good option. So I'm confused why now you're surprised that I went with the timer option over the tied-to-form-submission option.

ctsims commented on app/src/org/commcare/CommCareHeartbeatManager.java:
```diff
@@ -0,0 +1,309 @@
+
+    public void stopHeartbeatCommunications() {
+        if (heartbeatTimer != null) {
+            heartbeatTimer.cancel();
```
The "inside" of the heardbeat should likely have a flag set to cancel the poll, but we should also disconnect the "listener" from the heartbeat. Right now, the heartbeat thread will always be dependent on this class as long as it is running, which could result in a memory leak over it. This won't be needed if all of the state for the heartbeat processing can live inside of the internal heartbeat class, rather than the manger.

amstone326 replied:
Not following this so we should discuss in our sync.

amstone326 commented on app/src/org/commcare/CommCareHeartbeatManager.java:
```diff
@@ -0,0 +1,309 @@
+
+    private static final long ONE_DAY_IN_MS = 24 * 60 * 60 * 1000;
+
+    private static final String TEST_RESPONSE =
```
Where else would it go? It's a test response for the request that this class makes? (I'm also going to delete it once I'm able to test making the actual request and it works)

amstone326 commented on app/assets/locales/android_translatable_strings.txt:
```diff
@@ -19,6 +19,16 @@ updates.pinned.download=Downloading updates
+update.required.title=Update Required
+update.later.option=I'll update later.
+prompted.apk.update.info=There is a newer version of CommCare available, and it is recommended that you update now.
+forced.apk.update.info=There is a newer version of CommCare available. You must update before continuing to use CommCare.
```
Is what on our radar? Right now I just didn't design the non-forced updates to have any date associated with them, since I wasn't sure how we'd use it. I can add that if we want it just as something to display to the user in this message, but I was envisioning that by the time app builders would usually be setting an update prompt, they would most likely just want it to be done asap.

amstone326 commented on app/src/org/commcare/CommCareHeartbeatManager.java:
```diff
@@ -0,0 +1,309 @@
+                    //simulateRequestGettingStuck();
+                    //requestHeartbeat(currentUser);
+                    parseTestHeartbeatResponse();
+                } catch (SessionUnavailableException e) {
```
added

amstone326 commented on app/src/org/commcare/CommCareHeartbeatManager.java:
```diff
@@ -0,0 +1,309 @@
+ *
+ * Created by amstone326 on 4/13/17.
+ */
+public class CommCareHeartbeatManager {
```
done

amstone326 replied:
(this is now `HeartbeatLifecycleManager` and `HeartbeatRequester`)

amstone326 commented on app/src/org/commcare/CommCareHeartbeatManager.java:
```diff
@@ -0,0 +1,309 @@
+            }
+        };
+
+        heartbeatTimer = new Timer();
```
Addressed by always canceling the existing Timer, if it’s non-null, before instantiating a new one

ctsims commented:
Hey,

This is looking good based on what we talked about. I think one last judgement call we might need to make is how often to poll for doing the update, and I wonder whether the server team might have guidance on how often might be too often in high-load situations (IE: If they are rate limiting requests, retrying with higher frequency will keep the server overloaded).

amstone326 commented:
@ctsims Cool, so things I've fixed and added since you looked last night (encompassed in the most recent 6 commits, although 2 of those were a commit-and-revert):

- The place in `onServiceConnected` from which I was calling `initHeartbeatLifecycle()` was slightly off; should only be called when we have a user
- The way I was constructing a string from the response bytes was incorrect (discovered this while testing with the actual response from HQ; once that was fixed, the response was otherwise working as expected)
- Added a call to `initHeartbeatLifecycle()` immediately after form submission succeeds
- `initHeartbeatLifecycle()` was incorrectly always instantiating a new lifecycle manager object when called, which would have orphaned an existing one if there was one. Changed this to only instantiate the object if it's null.
- Decreased the frequency with which the TimerTask makes the request from once every 5 min to once every hour. I'll check in with Cal about the concern you raised, but this seemed like a good way to address it in the mean time. Especially now that we make the attempt after form submission (which is most likely to be when we succeed anyway), it seems fine to significantly reduce the rate of the random attempts.
- Added the IMEI and user id as default query params in `AndroidModernHttpRequester` (previously these were only being added to `HttpRequestGenerator.makeCaseFetchRequest`)

amstone326 commented:
Self-merging this so that we can code freeze. 

@ctsims if there are any changes you want based on my last comment, let me know later and they can go in during QA. I'm also planning on adding tests, and possibly UI changes based on feedback I get from Saket tomorrow.
